cli/command/manifest/push.go

```diff
@@ -0,0 +1,588 @@
+	if err != nil {
+		return nil, fmt.Errorf("Ping of V2 registry failed: %v", err)
+	}
+	if authConfig.RegistryToken != "" {
+		passThruTokenHandler := &existingTokenHandler{token: authConfig.RegistryToken}
+		modifiers = append(modifiers, auth.NewAuthorizer(challengeManager, passThruTokenHandler))
+	} else {
+		creds := registry.NewStaticCredentialStore(&authConfig)
+		tokenHandler := auth.NewTokenHandler(authTransport, creds, repoName, "*")
+		basicHandler := auth.NewBasicHandler(creds)
+		modifiers = append(modifiers, auth.NewAuthorizer(challengeManager, tokenHandler, basicHandler))
+	}
+	tr := transport.NewTransport(base, modifiers...)
+
+	httpClient := &http.Client{
+		Transport: tr,
+		// @TODO: Use the default (leave CheckRedirect nil), or write a generic one
+		// and put it somewhere? (There's one in docker/distribution/registry/client/repository.go)
+		// CheckRedirect: checkHTTPRedirect,
+	}
+	return httpClient, nil
+}
+
+func createManifestURLFromRef(targetRef reference.Named, urlBuilder *v2.URLBuilder) (string, error) {
+
+	manifestURL, err := urlBuilder.BuildManifestURL(targetRef)
+	if err != nil {
+		return "", fmt.Errorf("Failed to build manifest URL from target reference: %v", err)
+	}
+	return manifestURL, nil
+}
+
+func setupRepo(repoInfo *registry.RepositoryInfo) (registry.APIEndpoint, string, error) {
+	endpoint, err := selectPushEndpoint(repoInfo)
+	if err != nil {
+		return endpoint, "", err
+	}
+	logrus.Debugf("manifest: create: endpoint: %v", endpoint)
+	repoName := repoInfo.Name.Name()
+	// If endpoint does not support CanonicalName, use the RemoteName instead
+	if endpoint.TrimHostname {
+		repoName = reference.Path(repoInfo.Name)
+		logrus.Debugf("repoName: %v", repoName)
+	}
+	return endpoint, repoName, nil
+}
+
+func selectPushEndpoint(repoInfo *registry.RepositoryInfo) (registry.APIEndpoint, error) {
+	var err error
+
+	options := registry.ServiceOptions{}
+	// By default (unless deprecated), loopback (IPv4 at least...) is automatically added as an insecure registry.
+	options.InsecureRegistries, err = loadLocalInsecureRegistries()
+	if err != nil {
+		return registry.APIEndpoint{}, err
+	}
+	registryService := registry.NewService(options)
+	endpoints, err := registryService.LookupPushEndpoints(reference.Domain(repoInfo.Name))
+	if err != nil {
+		return registry.APIEndpoint{}, err
+	}
+	logrus.Debugf("manifest: potential push endpoints: %v\n", endpoints)
+	// Default to the highest priority endpoint to return
+	endpoint := endpoints[0]
+	if !repoInfo.Index.Secure {
+		for _, ep := range endpoints {
+			if ep.URL.Scheme == "http" {
+				endpoint = ep
+			}
+		}
+	}
+	return endpoint, nil
+}
+
+func loadLocalInsecureRegistries() ([]string, error) {
+	insecureRegistries := []string{}
+	// Check $HOME/.docker/config.json. There may be mismatches between what the user has in their
+	// local config and what the daemon they're talking to allows, but we can be okay with that.
+	userHome, err := homedir.GetStatic()
```

Comment:
The same function is used to [read the `config.json`](https://github.com/moby/moby/blob/34337db95d8eb85875509e341f452d8b555452b8/cli/cli.go#L23) for Docker Hub credentials, so I think this is fine. On Windows I do not have HOME set and it worked.